include/openthread/history_tracker.h

```diff
@@ -0,0 +1,211 @@
+ * average RSS of all fragment frames that form a received message and is only applicable for a RX message.
+ *
+ */
+typedef struct otHistoryTrackerMessageInfo
+{
+    uint16_t   mPayloadLength;       ///< IPv6 payload length (exclude IP6 header itself).
+    uint16_t   mNeighborRloc16;      ///< RLOC16 of neighbor which sent/received the msg (`0xfffe` if no RLOC16).
+    otSockAddr mSource;              ///< Source IPv6 address and port (if UDP/TCP)
+    otSockAddr mDestination;         ///< Destination IPv6 address and port (if UDP/TCP).
+    uint8_t    mIpProto;             ///< IP Protocol number (`OT_IP6_PROTO_*` enumeration).
+    uint8_t    mIcmp6Type;           ///< ICMP6 type if msg is ICMP6, zero otherwise (`OT_ICMP6_TYPE_*` enumeration).
+    int8_t     mAveRxRss;            ///< RSS of received message or OT_RADIO_INVALI_RSSI if not known.
+    bool       mLinkSecurity : 1;    ///< Indicates whether msg used link security.
+    bool       mTxSuccess : 1;       ///< Indicates TX success (e.g., ack received). Applicable for TX msg only.
+    uint8_t    mPriority : 2;        ///< Message priority (`OT_HISTORY_TRACKER_MSG_PRIORITY_*` enumeration).
+    bool       mRadioIeee802154 : 1; ///< Indicates whether msg was sent/received over a 15.4 radio link.
+    bool       mRadioTrelUdp6 : 1;   ///< Indicates whether msg was sent/received over a TREL radio link.
+} otHistoryTrackerMessageInfo;
+
+/**
+ * This function initializes an `otHistoryTrackerIterator`.
+ *
+ * An iterator MUST be initialized before it is used.
+ *
+ * An iterator can be initialized again to start from the beginning of the list.
+ *
+ * When iterating over entries in a list, to ensure the entry ages are consistent, the age is given relative to the
+ * time the iterator was initialized, i.e., the entry age is provided as the duration (in milliseconds) from the event
+ * (when entry was recorded) to the iterator initialization time.
+ *
+ * @param[in] aIterator  A pointer to the iterator to initialize (MUST NOT be NULL).
+ *
+ */
+void otHistoryTrackerInitIterator(otHistoryTrackerIterator *aIterator);
+
+/**
+ * This function iterates over the entries in the network info history list.
+ *
+ * @param[in]    aInstance   A pointer to the OpenThread instance.
+ * @param[inout] aIterator   A pointer to an iterator. MUST be initialized or the behavior is undefined.
+ * @param[out]   aEntryAge   A pointer to a variable to output the entry's age. MUST NOT be NULL.
+ *                           Age is provided as the duration (in milliseconds) from when entry was recorded to
+ *                           @p aIterator initialization time. It is set to `OT_HISTORY_TRACKER_MAX_AGE` for entries
+ *                           older than max age.
+ *
+ * @returns A pointer to `otHistoryTrackerNetworkInfo` entry or `NULL` if no more entries in the list.
+ *
+ */
+const otHistoryTrackerNetworkInfo *otHistoryTrackerIterateNetInfoHistory(otInstance *              aInstance,
```

Comment:
I do like and prefer the `Iterate` method name a lot better for these APIs (specially since we are are using an `Iterator`).

Funny thing is I used `GetNext()` name model initially when I was writing the code as a placeholder so that later I can find a better name 😄  (which I then changed to `Iterate{SomeList}()`).  `GetNext()` in my view is tied to linked-list.

----

I also found some interesting discussions (they point that it is  grammatically correct to say "iterate the list" compared to "iterate over the list" or "iterate through a list" (though the later two are used often) -> https://stackoverflow.com/questions/2715388/is-it-iterate-through-or-iterate-over-something

Reply:
We are using `otModuleGetNextSomething` a lot in OT APIs, but never used `otModuleIterateSomething` once. 

Even if `Iterate` may sounds better, we should keep wording in OT APIs consistent?

Reply:
I think using `Iterate` name here is better. It would be clear to OT API users what it means. We use `Iterate` in internal core in many places. It may be better to mirror the API names with the core internal names.

Reply:
I think `Iterate` makes more sense if we are using a C++ iterator, which is true for internal core code. 

But for C APIs without Iterator, `GetNext` seems more suitable?

Reply:
We are using an iterator in this API which we pass as a parameter.

I think we may be spending too much time discussing names and/or smaller nit/style topics which may not be very productive. Let's adopt the policy in PR reviews that we can suggest name changes, but then leave the decision to the person writing the code. I want us to avoid having a lot of back-and-forth on things like this where all options are fine and/or make little difference practically.